moonraker/components/wled.py

```diff
@@ -0,0 +1,502 @@
+from typing import (
+    TYPE_CHECKING,
+    Type,
+    List,
+    Any,
+    Optional,
+    Dict,
+    Coroutine,
+    Tuple,
+    Union,
+)
+
+if TYPE_CHECKING:
+    from confighelper import ConfigHelper
+    from websockets import WebRequest
+    from . import klippy_apis
+    APIComp = klippy_apis.KlippyAPI
+
+class ColorOrder(str, Enum):
+    RGB: str = "RGB"
+    RGBW: str = "RGBW"
+
+    def Elem_Size(self):
+        if self is ColorOrder.RGB:
+            return 3
+        return 4
+
+class OnOff(str, Enum):
+    on: str = "on"
+    off: str = "off"
+
+class Strip:
+    def __init__(self,
+                 name: str,
+                 color_order: ColorOrder,
+                 cfg: ConfigHelper):
+        self.server = cfg.get_server()
+        self.client = AsyncHTTPClient()
+        self.request_mutex = asyncio.Lock()
+
+        self.name = name
+        self.color_order = color_order
+
+        # Read the uri information
+        addr: str = cfg.get("address")
+        port: int = cfg.getint("port", 80)
+        protocol: str = cfg.get("protocol", "http")
+        self.url = f"{protocol}://{addr}:{port}/json"
+
+        self.timeout: float = cfg.getfloat("timeout", 2.)
+
+        self.initial_preset: int = cfg.getint("initial_preset", -1)
+        self.initial_red: float = cfg.getfloat("initial_red", 0.5)
+        self.initial_green: float = cfg.getfloat("initial_green", 0.5)
+        self.initial_blue: float = cfg.getfloat("initial_blue", 0.5)
+        self.initial_white: float = cfg.getfloat("initial_white", 0.5)
+        self.chain_count: int = cfg.getint("chain_count", 1)
+
+        self._chain_data = bytearray(self.chain_count * color_order.Elem_Size())
+
+        self.onoff = OnOff.off
+        self.preset = self.initial_preset
+
+    def get_strip_info(self) -> Dict[str, Any]:
+        return {
+            "strip": self.name,
+            "status": self.onoff,
+            "chain_count": self.chain_count,
+            "preset": self.preset,
+            "color_order": self.color_order,
+            "error": self.error_state,
+            "info": self.wled_info,
+            "effects": self.wled_effects,
+            "palettes": self.wled_palettes
+        }
+
+    async def initialize(self):
+        self.send_full_chain_data = True
+        self.onoff = OnOff.on
+        self.preset = self.initial_preset
+        if self.initial_preset >= 0:
+            self._update_color_data(self.initial_red,
+                                    self.initial_green,
+                                    self.initial_blue,
+                                    self.initial_white)
+            await self.wled_on(self.initial_preset)
+        else:
+            await self.set_wled(self.initial_red,
+                                self.initial_green,
+                                self.initial_blue,
+                                self.initial_white)
+
+    def _update_color_data(self, red, green, blue, white, index=None):
```

Comment:
The arguments and return value here should be annotated.

Reply:
Believed fixed